plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+		 g_warning("leave flash mode cmd error");
+		 return	FALSE;
+	}
+
+	if (!fu_ccgx_hpi_get_event (device, hpi_handle, HPI_REG_SECTION_DEV, &hpi_event, HPI_CMD_COMMAND_RESPONSE_TIME_MS ,error)) {
+		g_warning("leave flash mode resp error");
+		return FALSE;
+	}
+
+	if (hpi_event != CY_PD_RESP_SUCCESS) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "leave flash mode resp code error: 0x%x", hpi_event);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_write_flash:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @row_num row number
+ * @data flash row data
+ * @size size of row data
+ * @error: a #GError or %NULL
+ *
+ * Write Flash data to PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_write_flash (FuDevice *device, CyHPIHandle *hpi_handle, guint16 row_num, guint8 *data, guint16 size, GError **error)
+{
+	CyPDResp hpi_event = 0;
+	guint16 reg_addr = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+	g_return_val_if_fail (data != NULL, FALSE);
+	g_return_val_if_fail (size > 0, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event(device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
+
+	if (hpi_handle->hpi_addr_byte > 1) {
+		reg_addr = HPI_DEV_REG_FLASH_MEM;
+	} else {
+		reg_addr = CY_PD_REG_BOOTDATA_MEMEORY_ADDR;
+	}
+
+	/* write Data to memory */
+	if (!fu_ccgx_hpi_hpi_write_reg(device,hpi_handle, reg_addr, data, size,error)) {
+		g_warning("write data to memory error");
+		return FALSE;
+	}
+
+	g_usleep (HPI_CMD_FLASH_READ_WRITE_DELAY_US);
+
+	/* send write command */
+	if (!fu_ccgx_hpi_write_flash (device, hpi_handle, row_num, error)) {
+		g_warning("write flash cmd error");
+		return FALSE;
+	}
+
+	g_usleep (HPI_CMD_FLASH_READ_WRITE_DELAY_US); /* wait until flash is written */
+
+	if (!fu_ccgx_hpi_get_event (device, hpi_handle, HPI_REG_SECTION_DEV, &hpi_event, HPI_CMD_COMMAND_RESPONSE_TIME_MS ,error)) {
+		g_warning("write flash resp error");
+		return FALSE;
+	}
+
+	if (hpi_event != CY_PD_RESP_SUCCESS) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "write flash resp code error: 0x%x", hpi_event);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_read_flash:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @row_num row	number
+ * @data[out] flash row	data
+ * @size size of row data
+ * @error: a #GError or %NULL
+ *
+ * Read Flash data to PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_read_flash (FuDevice *device,
+			    CyHPIHandle *hpi_handle,
+			    guint16 row_num,
+			    guint8 *data,
+			    guint16 size,
+			    GError **error)
+{
+	guint16 reg_addr = 0;
+	CyPDResp hpi_event = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+	g_return_val_if_fail (data != NULL, FALSE);
+	g_return_val_if_fail (size > 0, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event(device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
```

Comment:
should return false if this fails

Reply:
OK. I will make fu_ccgx_hpi_clear_all_events get process GError and get return value when it has error.

Reply:
I fixed it.